src/ws/cockpitwebservice.c

```diff
@@ -2056,27 +2038,32 @@ cockpit_web_service_get_transport_init_message_finish (CockpitWebService *self,
   return init;
 }
 
-CockpitTransport *
-cockpit_web_service_find_transport (CockpitWebService *self,
-                                    const gchar *checksum)
+const gchar *
+cockpit_web_service_get_host (CockpitWebService *self,
+                              const gchar *checksum)
 {
-  CockpitSession *session;
-  GHashTableIter iter;
+  return g_hash_table_lookup (self->host_by_checksum, checksum);
+}
 
-  g_return_val_if_fail (COCKPIT_IS_WEB_SERVICE (self), NULL);
-  g_return_val_if_fail (checksum != NULL, NULL);
+const gchar *
+cockpit_web_service_get_checksum (CockpitWebService *self,
+                                  const gchar *host)
+{
+  return g_hash_table_lookup (self->checksum_by_host, host);
+}
 
-  /* Always check localhost first */
-  session = g_hash_table_lookup (self->sessions.by_host, "localhost");
-  if (session && session->checksum && g_str_equal (session->checksum, checksum))
-    return session->transport;
+void
+cockpit_web_service_set_host_checksum (CockpitWebService *self,
+                                       const gchar *host,
+                                       const gchar *checksum)
+{
+  const gchar *old_checksum = g_hash_table_lookup (self->checksum_by_host, host);
+  if (g_strcmp0 (checksum, old_checksum) == 0)
+    return;
 
-  g_hash_table_iter_init (&iter, self->sessions.by_transport);
-  while (g_hash_table_iter_next (&iter, NULL, (gpointer *)&session))
-    {
-      if (session->checksum && g_str_equal (session->checksum, checksum))
-        return session->transport;
-    }
+  if (old_checksum)
+    g_hash_table_remove (self->host_by_checksum, old_checksum);
```

Comment:
No, it's not redundant. If the old_checksum != new checksum and we just call replace, we will end up with 2 entries in host_by_checksum for host. Removing it explicitly is necessary.